File: Semester 4/Advanced Programming/RpcProtocolJava/ApplicationRepository/src/main/java/app/repository/CursaDbRepo.java
package app.repository;

import app.model.Cursa;
import jdk.vm.ci.meta.Local;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class CursaDbRepo implements ICursaRepo {
    private DbUtils dbutils;
    private Logger logger = LogManager.getLogger();
    public CursaDbRepo(Properties props){
        logger.info("Constructing CursaDbRepo with properties : {}",props);
        dbutils=new DbUtils(props);
    }
    @Override
    public void save(Cursa entity) {
        logger.traceEntry("saving Cursa {}",entity);
        Connection con = dbutils.getConnection();
        try(PreparedStatement stm = con.prepareStatement("insert into Curse values (?,?,?,?)"))
        {
            stm.setInt(1,entity.getId());
            stm.setString(2,entity.getDestinatie());
            stm.setString(3,entity.getDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")));
            stm.setInt(4,entity.getLocuriDisponibile());
            int result = stm.executeUpdate();
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Eroare la save in bd" + ex);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer key) {
        logger.traceEntry("deleting Cursa with {}",key);
        Connection con = dbutils.getConnection();
        try(PreparedStatement stm = con.prepareStatement("delete from Curse where id=?"))
        {
            stm.setInt(1,key);
            int result =stm.executeUpdate();
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Exceptie sql" + ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Cursa entity) {
        logger.traceEntry("Updating Cursa {}",entity);
        Connection con = dbutils.getConnection();
        try(PreparedStatement stm = con.prepareStatement("" +
                "update Curse set destinatie=?,date=?,locuri=? where id=?")){
            stm.setString(1,entity.getDestinatie());
            stm.setString(2,entity.getDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")));
            stm.setInt(3,entity.getLocuriDisponibile());
            stm.setInt(4,entity.getId());
            int result = stm.executeUpdate();
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println(ex);
        }
        logger.traceExit();

    }

    @Override
    public Cursa findOne(Integer key) {
        logger.traceEntry("Finding Cursa with {}",key);
        Connection con = dbutils.getConnection();
        try(PreparedStatement stm = con.prepareStatement("select * from Curse where id=?"))
        {
            stm.setInt(1,key);
            try(ResultSet result = stm.executeQuery()){
                if(result.next()){
                    int id = result.getInt("id");
                    String destinatie = result.getString("destinatie");
                    String r =result.getString("date");
                    LocalDateTime date = LocalDateTime.parse( result.getString("date"),DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
                    Integer nrLocuri = result.getInt("locuri");
                    return new Cursa(id,destinatie,date,nrLocuri);
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.out.println("Eroare sql la findOne " + ex);
        }
        logger.traceExit("Cursa with id {} not found",key);
        return null;
    }

    @Override
    public List<Cursa> findAll() {
        logger.traceEntry();
        Connection con = dbutils.getConnection();
        List<Cursa> curse = new ArrayList<>();
        try(PreparedStatement stm = con.prepareStatement("select * from Curse;")){
            try(ResultSet result  = stm.executeQuery()){
                while(result.next()){
                    int id = result.getInt("id");
                    String destinatie = result.getString("destinatie");
                    LocalDateTime date = LocalDateTime.parse( result.getString("date"),DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
                    Integer nrLocuri = result.getInt("locuri");
                    curse.add(new Cursa(id,destinatie,date,nrLocuri));
                }
            }
        } catch (SQLException e) {
            logger.error(e);
        }
        logger.traceExit(curse);
        return curse;
    }

    @Override
    public List<Cursa> findByDestinationAndDate(String destination, LocalDateTime date) {
        Connection con = dbutils.getConnection();
        List<Cursa> curse = new ArrayList<>();
        try{
            try(PreparedStatement stm = con.prepareStatement("select * from Curse where destinatie=? and date =?")){
                stm.setString(1,destination);
                stm.setString(2,date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")));
                try(ResultSet result = stm.executeQuery()){
                    while(result.next()){
                        int id = result.getInt("id");
                        String destinatie = result.getString("destinatie");
                        LocalDateTime date2 = LocalDateTime.parse( result.getString("date"),DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
                        Integer nrLocuri = result.getInt("locuri");
                        curse.add(new Cursa(id,destinatie,date2,nrLocuri));
                    }
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
        }
        try {
            con.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return curse;
    }
}

File: Semester 4/Advanced Programming/RpcProtocolC#/Repository/CursaDbRepo.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Domain;
using log4net;
namespace Repositories
{
    public class CursaDbRepo : ICursaRepo
    {
        public CursaDbRepo()
        {
            log.Info("Creating CursaDbRepo");
        }
        private static readonly ILog log = LogManager.GetLogger("CursaDbRepo");
        public void Delete(int id)
        {
            log.InfoFormat("Enter Delete id = {0}", id);
            IDbConnection con = DBUtils.getConnection();

            using (var cmd = con.CreateCommand())
            {
                cmd.CommandText = "delete from Curse where id = @id";
                IDbDataParameter param = cmd.CreateParameter();
                param.ParameterName = "@id";
                param.Value = id;
                cmd.Parameters.Add(param);
                int result = cmd.ExecuteNonQuery();
                con.Close();
            }
            log.InfoFormat("Exit Delete");
        }

        public IEnumerable<Cursa> FindAll()
        {
            log.InfoFormat("Enter findAll()");
            IList<Cursa> curse = new List<Cursa>();
            var con = DBUtils.getConnection();
            using (var cmd = con.CreateCommand())
            {

                cmd.CommandText = "Select * from Curse";

                using (var dataReader = cmd.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        int id2 = dataReader.GetInt32(0);
                        string destinatie = dataReader.GetString(1);
                        DateTime date = dataReader.GetDateTime(2);
                        int locuri = dataReader.GetInt32(3);
                        Cursa cursa = new Cursa(id2, destinatie, date, locuri);
                        curse.Add(cursa);
                    }
                    con.Close();
                }
            }
            log.InfoFormat("Exit findAll with :{0}", curse);
            return curse;
        }

        public IEnumerable<Cursa> FindByDestinationAndDate(string destination, DateTime date)
        {
            log.InfoFormat("Enter findByDestinationAndDate with destination={0} ; date={1}", destination, date);
            IList<Cursa> curse = new List<Cursa>();
            var con = DBUtils.getConnection();
            using (var cmd = con.CreateCommand())
            {
                cmd.CommandText = "Select * from Curse where destinatie = @destination and date = @date";
                IDbDataParameter destParam = cmd.CreateParameter();
                destParam.ParameterName = "@destination";
                destParam.Value = destination;
                IDbDataParameter dateParam = cmd.CreateParameter();
                dateParam.ParameterName = "@date";
                dateParam.Value = date;
                cmd.Parameters.Add(destParam);
                cmd.Parameters.Add(dateParam);

                using (var dataReader = cmd.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        int id2 = dataReader.GetInt32(0);
                        string destinatie = dataReader.GetString(1);
                        DateTime date2 = dataReader.GetDateTime(2);
                        int locuri = dataReader.GetInt32(3);
                        Cursa cursa = new Cursa(id2, destinatie, date, locuri);
                        curse.Add(cursa);
                    }
                    con.Close();
                }
            }
            log.InfoFormat("Exit findByDestinationAndDate with :{0}", curse);
            return curse;
        }

        public Cursa FindOne(int id)
        {
            log.InfoFormat("Enter FIndOne with key = {0}", id);
            var con = DBUtils.getConnection();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select * from Curse where id=@id";
                IDbDataParameter param = comm.CreateParameter();
                param.ParameterName = "@id";
                param.Value = id;
                comm.Parameters.Add(param);
                using (var dataReader = comm.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        int id2 = dataReader.GetInt32(0);
                        string destinatie = dataReader.GetString(1);
                        DateTime date = dataReader.GetDateTime(2);
                        int locuri = dataReader.GetInt32(3);
                        Cursa cursa = new Cursa(id2, destinatie, date, locuri);
                        log.InfoFormat("Exit findOne with :{0}", cursa);
                        con.Close();
                        return cursa;
                    }
                }
            }
            log.InfoFormat("Exit findOne with :{0}", null);
            return null;
        }

        public void Save(Cursa entity)
        {
            var con = DBUtils.getConnection();
            using (var cmd = con.CreateCommand())
            {
                cmd.CommandText = "insert into Curse values(@id,@destinatie,@date,@locuri)";
                var idParam = cmd.CreateParameter();
                var destParam = cmd.CreateParameter();
                var dateParam = cmd.CreateParameter();
                var locuriParam = cmd.CreateParameter();

                idParam.ParameterName = "@id";
                destParam.ParameterName = "@destinatie";
                dateParam.ParameterName = "@date";
                locuriParam.ParameterName = "@locuri";

                idParam.Value = entity.id;
                destParam.Value = entity.destinatie;
                dateParam.Value = entity.date;
                locuriParam.Value = entity.locuriDisponibile;

                cmd.Parameters.Add(idParam);
                cmd.Parameters.Add(destParam);
                cmd.Parameters.Add(dateParam);
                cmd.Parameters.Add(locuriParam);

                int result = cmd.ExecuteNonQuery();
                con.Close();
            }
        }

        public void Update(Cursa entity)
        {
            var con = DBUtils.getConnection();
            using (var cmd = con.CreateCommand())
            {
                cmd.CommandText = "update Curse set destinatie=@dest,date=@date,locuri=@locuri where id=@id";
                var destParam = cmd.CreateParameter();
                var dateParam = cmd.CreateParameter();
                var locuriParam = cmd.CreateParameter();
                var idParam = cmd.CreateParameter();

                destParam.ParameterName = "@dest";
                dateParam.ParameterName = "@date";
                locuriParam.ParameterName = "@locuri";
                idParam.ParameterName = "@id";

                destParam.Value = entity.destinatie;
                dateParam.Value = entity.date;
                locuriParam.Value = entity.locuriDisponibile;
                idParam.Value = entity.id;

                cmd.Parameters.Add(destParam);
                cmd.Parameters.Add(dateParam);
                cmd.Parameters.Add(locuriParam);
                cmd.Parameters.Add(idParam);

                int result = cmd.ExecuteNonQuery();
                con.Close();
            }
        }

    }
}
